### PR TITLE
chore: fix mypy pre-commit hook

### PR DESCRIPTION
**Pull request by nejch**
Closes https://github.com/python-gitlab/python-gitlab/issues/1539.

Additionally, since mypy 0.910 doesn't bundle stubs anymore we need to track them for pre-commit's own virtualenv :/ See https://github.com/pre-commit/pre-commit/issues/730

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,6 +3,28 @@ profile = "black"
 multi_line_output = 3
 order_by_type = false
 
+[tool.mypy]
```
Comment (JohnVillalovos) [inline pyproject.toml:6]: So we also have `.mypy.ini`

Why this change here?

Comment (nejch) [reply pyproject.toml:6]: Sorry, I forgot to mention. While upgrading to mypy 0.91 and looking at docs I found that it now supports `pyproject.toml` so I thought I'd consolidate more tool config in there. I've actually deleted `.mypy.ini` in this PR. But I just realized I left out the comments from the previous config, I'll amend that.

Comment (JohnVillalovos) [reply pyproject.toml:6]: Ah. I somehow missed noticing the deletion of the file.

File: pyproject.toml [hunk 2]
```diff
@@ -3,6 +3,28 @@ profile = "black"
...
+    "tests.functional.*",
+    "tests.functional.api.*",
+    "tests.unit.*"
+]
+ignore_errors = true
+
+[[tool.mypy.overrides]] # Overrides to negate above patterns
+module = [
+    "gitlab.v4.objects.projects"
+]
```
Comment (antti-mikael) [inline pyproject.toml:25]: The overrides `module` listing does not match the previous `files` definition from `.mypy.ini`. Is this intentional?

Comment (nejch) [reply pyproject.toml:25]: Yes, since `pre-commit` explicitly passes files to mypy (including sometimes `docs/`, `tests/`, I'm unsure why), I had to add some files outside `gitlab/` to ignores as well. Since it only supports wildcards, I then had to negate the `gitlab.v4.objects.*` ignore pattern with this one, since it's already typed.

File: tox.ini [hunk 3]
```diff
@@ -41,7 +41,7 @@ basepython = python3
 envdir={toxworkdir}/lint
 deps = -r{toxinidir}/requirements-lint.txt
 commands =
-  mypy {posargs}
+  mypy {posargs} .
```
Comment (javatarz) [inline tox.ini:44]: This might break when `tox -e mypy -- tests/unit/test_cli.py` is run for a single file/sub-dir check.

Comment (javatarz) [inline tox.ini:44]: The `.` causes failures when `tox` is used to run `mypy` on a single file with the command `tox -e mypy -- tests/unit/test_cli.py`

[the remark this answers is not in the document]
Comment (javatarz) [reply tox.ini:44]: I've looked at `mypy` docs and issues and there seems to be no way to address failures when the same file shows up twice. Don't think we can fix this one.

**Review by javatarz — APPROVED** [review]

File: tox.ini (hunk 3 above)
[the remark this answers is not in the document]
Comment (nejch) [reply tox.ini:44]: Thanks for that! I can actually move this to just be `files = "."` inside `pyproject.toml` (`gitlab/` would be better but pre-commit wouldn't care anyway), and then it'll just override it and should work.

**Comment by javatarz** [comment]
Can we please get this merged? The pre-commit hooks are broken and the `mypy` stage blocks commits every time.

The workaround I've been using is to disable `mypy` locally (via a commit) and then remove the commit before pushing. This is painful.

File: tox.ini (hunk 3 above)
[the remark this answers is not in the document]
Comment (nejch) [reply tox.ini]: Done.

**Comment by nejch** [comment]
> Can we please get this merged? The pre-commit hooks are broken and the `mypy` stage blocks commits every time.
> 
> The workaround I've been using is to disable `mypy` locally (via a commit) and then remove the commit before pushing. This is painful.

Sorry, I know this was really annoying!

@JohnVillalovos I'm back from the dead :laughing: would you be able to take a quick look/merge?

**Comment by JohnVillalovos** [comment]
Thanks @nejch !
